### PR TITLE
Extend ESP32 CAN bit rates /bus speed support

### PR DESCRIPTION
**Pull request by mkaiser**
# What does this implement/fix?

The internal CAN bus controller of newer ESP32 variants support more bit rates than the original ESP32. 

This commit adds support for all supported CAN bus speeds listed in [ESP API Reference](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/twai.html#driver-configuration)

Added bit rate support for ESP32 S2, S3, C3 and H2
  1KBPS
  5KBPS
  10KBPS
  12K5BPS
  16KBPS
  20KBPS
  800KBPS    (including original ESP32)

Tested with my heatpump "Tecalor TTF 07 cool" on an ESP32-C3, CAN transceiver "TI SN65HVD230", CAN bit_rate @ 20KBPS

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3142

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

esphome:
  name: test-esp32-can-s3
  friendly_name: test-esp32-can-s3

esp32:
  board: esp32dev
  #board: esp32-s2-saola-1
  #board: esp32-s3-devkitc-1
  #board: esp32-c3-devkitm-1
   
  framework:
    type: arduino

canbus:
  - platform: esp32_can
    tx_pin: GPIO5
    rx_pin: GPIO4
    can_id: 4
    bit_rate: 125KBPS
    on_frame:
    - can_id:      0  # listen to all messages
      can_id_mask: 0
      use_extended_id: true
      then:
        - lambda: |-
            ESP_LOGD("can_id", "Received CAN message with ID: %x", can_id);        


# supported bit_rates for newer ESP32 variants
#1KBPS 
#5KBPS 
#10KBPS
#12K5BPS
#16KBPS
#20KBPS
#25KBPS
#50KBPS
#100KBPS
#125KBPS
#250KBPS
#500KBPS
#800KBPS
#1000KBPS

#unsupported bit_rates for ALL ESP32
#31K25BPS	 	 	 	 	 	 
#33KBPS	 	 	 	 	 	 
#40KBPS
#80KBPS	 	 	 	 	 	 
#83K38BPS	 	 	 	 	 	 
#95KBPS

# example: prevented by schema: 
# board: esp32dev
# bit_rate: 1kbps
```

## Checklist:
  - [ x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mkaiser** [comment]
If sources are okay - could you make a **squashed PR** when merging (had code formatting issues...)

File: esphome/components/esp32_can/canbus.py [hunk 1]
```diff
@@ -5,26 +5,74 @@
...
+}
+
+
+def validate_bit_rate(value):
+    variant = get_esp32_variant()
+    if variant not in CAN_SPEEDS:
+        raise cv.Invalid(f"{variant} is not supported by component {esp32_can_ns}")
+    if value not in CAN_SPEEDS[variant]:
+        raise cv.Invalid(f"Bit rate {value} is not supported on {variant}")
+    return value
```
Comment (jesserockz) [inline esphome/components/esp32_can/canbus.py]:
```suggestion
def validate_bit_rate():
    variant = get_esp32_variant()
    def validator(value):
	    if variant not in CAN_SPEEDS:
	        raise cv.Invalid(f"{variant} is not supported by component {esp32_can_ns}")
	    return cv.enum(CAN_SPEEDS[variant], upper=True)
	return validator
```

File: esphome/components/esp32_can/canbus.py [hunk 2]
```diff
@@ -5,26 +5,74 @@
...
+
+
 CONFIG_SCHEMA = canbus.CANBUS_SCHEMA.extend(
     {
         cv.GenerateID(): cv.declare_id(esp32_can),
-        cv.Optional(CONF_BIT_RATE, default="125KBPS"): cv.enum(CAN_SPEEDS, upper=True),
+        cv.Optional(CONF_BIT_RATE, default="125KBPS"): lambda value: validate_bit_rate(
+            value.upper()
+        ),
```
Comment (jesserockz) [inline esphome/components/esp32_can/canbus.py]:
```suggestion
        cv.Optional(CONF_BIT_RATE, default="125KBPS"): validate_bit_rate,
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
This is more how we do the validations

**Comment by mkaiser** [comment]
> This is more how we do the validations

sorry, I could not get it to work with the def validator(value) function. 

The committed version works, but I guess it is not what you intended :/

Opened a discord thread for this. maybe its just a small python syntax issue here. 

https://discord.com/channels/429907082951524364/1147632418472341544

**Comment by mkaiser** [comment]
[jesserockz](https://github.com/jesserockz) Thank you for the clarification on the def validator (discord)!

From my side, the PR is now ready :)

File: esphome/components/esp32_can/canbus.py [hunk 3]
```diff
@@ -5,26 +5,73 @@
...
+
+
+def validate_bit_rate(value):
+    variant = get_esp32_variant()
+    if variant not in CAN_SPEEDS:
+        raise cv.Invalid(f"{variant} is not supported by component {esp32_can_ns}")
+    value = value.upper()
+    if value not in CAN_SPEEDS[variant]:
+        raise cv.Invalid(f"Bit rate {value} is not supported on {variant}")
+    return value
```
Comment (jesserockz) [inline esphome/components/esp32_can/canbus.py]:
```suggestion
    return cv.enum(CAN_SPEEDS[variant])(value)
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**